core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/StopStartUpscaleGetRecoveryCandidatesHandler.java

```diff
@@ -0,0 +1,143 @@
+                    collectStartedInstancesFromCloudProvider(connector, ac, request.getAllInstancesInHostGroup())
+                            .stream()
+                            .map(CloudInstance::getInstanceId).collect(Collectors.toSet()));
+
+            ClusterHealthService clusterHealthService = clusterApiConnectors.getConnector(stack).clusterHealthService();
+
+            DetailedHostStatuses detailedHostStatuses = clusterHealthService.getDetailedHostStatuses();
+
+            List<InstanceMetadataView> instanceMetadataWithServicesNotRunning = collectRecoveryCandidates(startedInstances, detailedHostStatuses);
+
+            List<CloudInstance> instancesWithServicesNotRunning = instanceMetaDataToCloudInstanceConverter.convert(instanceMetadataWithServicesNotRunning,
+                    stack.getStack());
+
+            StopStartUpscaleGetRecoveryCandidatesResult result =
+                    new StopStartUpscaleGetRecoveryCandidatesResult(request.getResourceId(), request, instancesWithServicesNotRunning,
+                            request.getAdjustment(), request.getHostGroupName());
+            notify(result, event);
+        } catch (Exception e) {
+            String errorMessage = String.format("Failed to collect upscale recovery candidates for stack: %s, proceeding with subsequent flow.",
+                    request.getResourceId());
+            LOGGER.error(errorMessage, e);
+            StopStartUpscaleGetRecoveryCandidatesResult result = new StopStartUpscaleGetRecoveryCandidatesResult(request.getResourceId(), request,
+                    Collections.emptyList(), request.getAdjustment(), request.getHostGroupName());
+            notify(result, event);
+        }
+    }
+
+    private List<CloudInstance> collectStartedInstancesFromCloudProvider(CloudConnector<?> connector, AuthenticatedContext ac,
+            List<CloudInstance> cloudInstances) {
+        List<CloudVmInstanceStatus> vmInstanceStatuses = connector.instances().checkWithoutRetry(ac, cloudInstances);
+        return vmInstanceStatuses.stream()
+                .filter(vm -> InstanceStatus.STARTED.equals(vm.getStatus()))
+                .map(CloudVmInstanceStatus::getCloudInstance)
+                .collect(Collectors.toList());
+    }
+
+    private AuthenticatedContext getAuthenticatedContext(StopStartUpscaleGetRecoveryCandidatesRequest request, CloudContext cloudContext,
+            CloudConnector<?> connector) {
+        return connector.authentication().authenticate(cloudContext, request.getCloudCredential());
+    }
+
+    private List<InstanceMetadataView> collectRecoveryCandidates(List<InstanceMetadataView> startedInstances, DetailedHostStatuses detailedHostStatuses) {
+        Predicate<InstanceMetadataView> instanceHealthCheck = im -> {
+            HostName hostName = hostName(im.getDiscoveryFQDN());
+            return detailedHostStatuses.isHostDecommissioned(hostName)
+                    || !detailedHostStatuses.isHostHealthy(hostName)
```

Comment:
The serviceCheck and hostCheck being clubbed together likely needs to be split. Also don't see CONCERNING - do we not need to handle that explicitly, or is this implementation of detailedHostStatuses taking care of it while setting HEALTHY/UNHEALTHY.